homeassistant/helpers/script.py

```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+
+    async def _async_event_step(self):
+        """Fire an event."""
+        self._script.last_action = self._action.get(
+            CONF_ALIAS, self._action[CONF_EVENT]
+        )
+        self._log("Executing step %s", self._script.last_action)
+        event_data = dict(self._action.get(CONF_EVENT_DATA, {}))
+        if CONF_EVENT_DATA_TEMPLATE in self._action:
+            try:
+                event_data.update(
+                    template.render_complex(
+                        self._action[CONF_EVENT_DATA_TEMPLATE], self._variables
+                    )
+                )
+            except exceptions.TemplateError as ex:
+                self._log(
+                    "Error rendering event data template: %s", ex, level=logging.ERROR
+                )
+
+        self._hass.bus.async_fire(
+            self._action[CONF_EVENT], event_data, context=self._context
+        )
+
+    async def _async_condition_step(self):
+        """Test if condition is matching."""
+        config_cache_key = frozenset((k, str(v)) for k, v in self._action.items())
+        config = self._config_cache.get(config_cache_key)
+        if not config:
+            config = await condition.async_from_config(self._hass, self._action, False)
+            self._config_cache[config_cache_key] = config
+
+        self._script.last_action = self._action.get(
+            CONF_ALIAS, self._action[CONF_CONDITION]
+        )
+        check = config(self._hass, self._variables)
+        self._log("Test condition %s: %s", self._script.last_action, check)
+        if not check:
+            raise _StopScript
+
+    def _log(self, msg, *args, level=logging.INFO):
+        self._script._log(msg, *args, level=level)  # pylint: disable=protected-access
+
+    def _raise(self, msg, *args, exception=None):
+        # pylint: disable=protected-access
+        self._script._raise(msg, *args, exception=exception)
+
+
+class _ScriptRun(_ScriptRunBase):
+    """Manage Script sequence run."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        script: "Script",
+        variables: Optional[Sequence],
+        context: Optional[Context],
+    ) -> None:
+        super().__init__(hass, script, variables, context)
+        self._stop = asyncio.Event()
+        self._stopped = asyncio.Event()
+
+    async def _async_run(self) -> None:
+        self._log("Running script")
+        try:
+            for self._step, self._action in enumerate(self._script.sequence):
+                if self._stop.is_set():
+                    break
+                await self._async_step()
+        except _StopScript:
+            pass
+        finally:
+            if not self._stop.is_set():
+                self._changed()
+            self._script._runs.remove(self)  # pylint: disable=protected-access
+            self._stopped.set()
+
+    async def async_stop(self) -> None:
+        """Stop script run."""
+        self._stop.set()
+        await self._stopped.wait()
```

Comment:
This function should still cancel any ongoing delay/wait for template to be changed. Maybe we can have a `self._stoppable_task` variable that this method will cancel if it's set ?

Reply:
It does stop them. Both `_async_delay_step` & `_async_wait_template_step` also monitor the `self._stop` Event.

Reply:
Nevermind this, just noticed your `async_delay_step` implementation!